src/transform/src/threshold_elision.rs

```diff
@@ -32,32 +32,58 @@ impl crate::Transform for ThresholdElision {
         relation: &mut MirRelationExpr,
         _: TransformArgs,
     ) -> Result<(), crate::TransformError> {
-        relation.try_visit_mut_post(&mut |e| self.action(e))
+        let mut visitor = ThresholdElisionAction::default();
+        relation.visit_mut(&mut visitor).map_err(From::from)
     }
 }
 
-impl ThresholdElision {
+#[derive(Default)]
+struct ThresholdElisionAction {
+    non_negative: NonNegative,
+}
+
+impl VisitorMut<MirRelationExpr> for ThresholdElisionAction {
+    fn pre_visit(&mut self, expr: &mut MirRelationExpr) {
+        self.non_negative.pre_visit(expr);
+    }
+
+    fn post_visit(&mut self, expr: &mut MirRelationExpr) {
+        // Derive non_negative attributes before attempting the transform,
+        // as the latter depens on the former.
+        self.non_negative.post_visit(expr);
+        self.action(expr);
+    }
+}
+
+impl ThresholdElisionAction {
     /// Remove Threshold operators with no effect.
-    pub fn action(&self, relation: &mut MirRelationExpr) -> Result<(), crate::TransformError> {
-        if let MirRelationExpr::Threshold { input } = relation {
+    pub fn action(&mut self, expr: &mut MirRelationExpr) {
+        if let MirRelationExpr::Threshold { input } = expr {
             // We look for the pattern `Union { base, Negate(Subset(base)) }`.
             let mut should_replace = false;
             if let MirRelationExpr::Union { base, inputs } = &mut **input {
                 if inputs.len() == 1 {
                     if let MirRelationExpr::Negate { input } = &inputs[0] {
-                        let mut safe_lets = HashSet::default();
-                        if non_negative(base, &mut safe_lets) && lhs_superset_of_rhs(base, &*input)
-                        {
+                        // This is somewhat convoluted way to access the non_negative result for the base:
+                        // - the Threshold is at position n - 1,
+                        // - the Union (i.e., the Threshold input) is n - 2,
+                        // - the Union inputs are at positions n - 2 - 1 .. n - 2 - inputs.len(),
+                        // - the Union base is at position n - 2 - inputs.len() - 1 = n - 3 - inputs.len()
```

Comment:
As discussed on zoom, this index might be off if the subtrees are not simple. This is because you want to jump over one or more inputs, and results for an input subtree are in a contiguous interval of `results`, and the length of this interval depends on the size of the substree, i.e., they are 1 only if the subtree is simple.

Reply:
Ouch! That was a tricky one, I owe you some :beers:...